tst/EnergyPlus/unit/ElectricPowerServiceManager.unit.cc

```diff
@@ -1230,3 +1230,103 @@ TEST_F(EnergyPlusFixture, Battery_checkUserEfficiencyInputTest)
+        "  0,                       !- Minimum Value of x",
+        "  1,                       !- Maximum Value of x",
+        "  -100,                    !- Minimum Curve Output",
+        "  100,                     !- Maximum Curve Output",
+        "  Dimensionless,           !- Input Unit Type for x",
+        "  Dimensionless;           !- Output Unit Type",
+
+    });
+    ASSERT_TRUE(process_idf(idf_objects));
+
+    Curve::GetCurveInput(*state);
+    state->dataCurveManager->GetCurvesInputFlag = false;
+
+    // Test 1: Curves which do not produce errors
+    e0c = 13.0;
+    e0d = 11.0;
+    chaCurveIndex = 1;
+    disCurveIndex = 2;
+
+    checkChargeDischargeVoltageCurves(*state, "Battery for Test 1", e0c, e0d, chaCurveIndex, disCurveIndex);
+
+    EXPECT_TRUE(compare_err_stream(""));
+
+    // Test 2: Curves from ShopwithPVBattery.idf as of 7/2023
+    e0c = 12.6;
+    e0d = 12.4;
+    chaCurveIndex = 3;
+    disCurveIndex = 4;
+
+    checkChargeDischargeVoltageCurves(*state, "Battery for Test 2", e0c, e0d, chaCurveIndex, disCurveIndex);
+
+    std::string const error_string2 = delimited_string({
+        "   ** Warning ** Kinetic Battery Model: Battery for Test 2 has a charging/discharging voltage curve conflict.",
+        "   **   ~~~   ** Discharging voltage is higher than charging voltage which may potentially lead to an imbalance in the stored energy.",
+        "   **   ~~~   ** Check the charging and discharging curves to make sure that the charging voltage is greater than discharging.",
+        "   **   ~~~   ** Also check the charging and discharging energy outputs to find any discrepancies.",
+        "   **   ~~~   ** Charged fraction = 0.0, Charging voltage = 12.400 V, Discharging voltage = 12.591 V",
+        "   **   ~~~   ** Charged fraction = 0.1, Charging voltage = 12.401 V, Discharging voltage = 12.592 V",
+        "   **   ~~~   ** Charged fraction = 0.2, Charging voltage = 12.402 V, Discharging voltage = 12.593 V",
+        "   **   ~~~   ** Charged fraction = 0.3, Charging voltage = 12.403 V, Discharging voltage = 12.594 V",
+        "   **   ~~~   ** Charged fraction = 0.4, Charging voltage = 12.404 V, Discharging voltage = 12.595 V",
+        "   **   ~~~   ** Charged fraction = 0.5, Charging voltage = 12.405 V, Discharging voltage = 12.595 V",
+        "   **   ~~~   ** Charged fraction = 0.6, Charging voltage = 12.406 V, Discharging voltage = 12.596 V",
+        "   **   ~~~   ** Charged fraction = 0.7, Charging voltage = 12.407 V, Discharging voltage = 12.597 V",
+        "   **   ~~~   ** Charged fraction = 0.8, Charging voltage = 12.408 V, Discharging voltage = 12.598 V",
+        "   **   ~~~   ** Charged fraction = 0.9, Charging voltage = 12.409 V, Discharging voltage = 12.599 V",
+        "   **   ~~~   ** Charged fraction = 1.0, Charging voltage = 12.410 V, Discharging voltage = 12.600 V",
+    });
+    EXPECT_TRUE(compare_err_stream(error_string2, true));
+}
```

Comment:
And more compact and descriptive message.

Reply:
:+1: